### ISSUE TITLE
[FR] Better support for ndarrays and tensors

### ISSUE THREAD

**Issue by ntakouris** [mlflow/mlflow#5739]
**Please fill in this feature request template to ensure a timely and thorough response.**

## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [  x ] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

Many times, you log metrics such as loss per step (tensors with gradients) or other numpy arrays (ndarrays). It would be very useful if there was built-in support to:

**For params**
- Extract single (`.item()`) or multiple values from ndarrays ([tolist](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html) ?)

**For metrics**
- [Detach torch tensors from the gradient graph](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html) - pytorch lightning does this too
- Move tensors from gpu to cpu, convert to numpy if needed (ex. to convert to list)
- Extract values from single-element tensors or ndarrays, raise if resulting value is not supported (not float?)

Adding those few simple features, can make the usage much more intuitive and simpler, for deep learning engineers. Here's an example:

```py
anchors = np.array([[16], [32], [64], [128], [256]], dtype=np.int32)
x = np.array([1.0])

# in your train on batch func
loss = ...
```

- Before
```py
mlflow.log_params({'anchors': anchors.tolist(), 'x': x.item(), 'loss': loss.detach().item()/.tolist()})
````

- After
```py
mlflow.log_params({'anchors': anchors, 'x': x, 'loss': loss})
````

### What component(s), interfaces, languages, and integrations does this feature affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ x ] `area/server-infra`: MLflow Tracking server backend
- [ x ] `area/tracking`: Tracking Service, tracking client APIs, autologging

**Comment by WeichenXu123** [comment]
Good proposal !

1. We can support array type params, but we do not support array type metric.
2. We can support automatically call detach() / item() / tolist() , etc. methods.

@ntakouris Are you interested in contributing this feature ? :)

**Comment by ntakouris** [comment]
@WeichenXu123 yes, but please provide some guidance regarding files and constructs that need change.
